op-node-derivers.md

```diff
@@ -0,0 +1,554 @@
+A test-runner would simply iterate through a list of encoded events, applied to a deriver,
+on top of some pre-state, and then assert some post-state properties.
+
+### Alt-clients: Rust / Java
+
+Typed events and event-switches are very natural in Rust and Java, and even more common than in Go.
+Alternative rollup node implementations should be able to adopt this pattern.
+
+### Migration path
+
+This one of the more involved refactors, that we cannot execute all at once in op-node.
+This is a proposal for a 3-phase migration:
+
+#### 1) Derivation/Engine decoupling
+
+The derivation pipeline currently is coupled too tightly to the Engine.
+In particular, the `EngineQueue` should be removed,
+to allow the attributes-generation and unsafe-block processing to be truly independent processes.
+
+The derivation is somewhat statefully connected to the engine however:
+- Block-attributes result in later promotion to a "safe" head
+- Not all attributes are equal: those generated from span-batches result in unsafe blocks, until the last in the span is processed.
+- The find-sync-start routine is coupled to the `EngineQueue` `Reset()`
+
+We need a first step that addresses this, so the derivation process that generates the block attributes
+is shareable with the new driver design.
+
+#### 2) Incremental `OnEvent` adoption
+
+We can incrementally change the Driver state-loop sub-processes to fit the `OnEvent` signature,
+without changing the synchronous behavior of the state loop.
+This way the node behaves the same, and there is no risk of new concurrency bugs.
+
+#### 3) Swapping the driver
+
+Once all the state-loop sub-processes are really functional `Deriver`s,
+we can swap the `Driver` with one that implements the resource-based scheduling.
+This can be implemented alongside the current driver, and swapped by feature-flag in the `Driver` constructor,
+to introduce the concurrency-affecting part of the refactor more gracefully.
+
+An illustrative example of this approach can be found in the draft of the `RepeatCond` in
+[monorepo PR 10051](https://github.com/ethereum-optimism/optimism/pull/10051):
+
+We can make the `Driver` itself the root deriver, which accepts external inputs through public methods,
+and turns them into events, executing them with the relevant resource-specific worker go routine.
+
+# Risks & Uncertainties
+
+<!-- *An overview of what could go wrong. Also any open questions that need more work to resolve.* -->
+
+## Introduction of concurrency bugs
+
+When running derivers on different threads, there is state that may be shared between them, and needs to be safe.
+Deadlocks are especially important to avoid.
+Derivers should not acquire one resource before acquiring the other one when other derivers also do so.
+`ctx`-timeouts might unblock this, but the stall of locking up like this is already bad.
+Instead, derivers should design such that they communicate contentious work to a shared type of `Resource` worker.
```

Comment:
The `Resource` worker is essentially a router/prioritizer here? This helps explain my earlier confusion about the relationship between Resources and Workers a bit. 

With regards to resources: Without something in the design to prevent this, it's inevitable that we'd eventually have some concurrency error(s) over time.

I suppose the root deriver would mostly be calling to high level router/prioritizer workers, and the true implementation `Deriver` workers are leaf nodes of the structure. Then managing contentious resources is just a matter of router implementation.

Are there any cases where work has multiple contentious Resources that aren't managed by just a single `Resource worker`?

Reply:
Answered in aggregate resource/event comment below.